### ISSUE TITLE
Consider making errors always detailed

### ISSUE THREAD

**Issue by roji** [dotnet/efcore#23038]
We have the EnableDetailedErrors options, which isn't the default in order to avoid an try/catch that is otherwise unneeded. It's unlikely that this try/catch could have any sort of meaningful perf impact, so we should look into removing the option and making errors always detailed.

**Comment by ajcvickers** [comment]
Keep in mind that this is one try/catch _per value read from the reader_. So if you're reading 1000 rows with 20 mapped columns in each row, that would be additional 20,000 try/catch blocks executed.

**Comment by roji** [comment]
Yeah, understood. I definitely won't be doing anything here unless it's clearly proven to not degrade performance.

For reference, [here's the ASM-level difference](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAMIR8AB14AbGFADKMgG68wMXAG4aNYgGYmpBkJoBvGgzMMA2gFkYGABYQAJgElxkgBQ37T12MkB5MT4ILlwWADkIZy5JXi44gHMASgBdU3NmJCYUBgB1Xm8ODAAVKABPIWwMMDt3JPSzE2pzFoYAMQgIOrVW8waGAF9+/szsvIK7UoqqmrqGfqbeswxy/pbFpZaOrqSNZqWh/d6wGbs18w3ew5brvqPLLwcXN09bJ98AoN4QsMjo2PiXGSaXuo2IOW2dQW5zMzAAnO4AERtfz+RG7fqHAZAA===) of a non-inlinable method with and without a try/catch:

**Comment by ajcvickers** [comment]
@roji I was thinking about this last night. It may be possible to track which column is currently being read by incrementing a counter each read, for example. This could allow us to detect which read failed without needing to try catch each one. (I don't know whether this will have more/less overhead. than try/catch.)

**Comment by roji** [comment]
One last thought - if we end up keeping the option (because try/catch does have some noticeable overhead), we may still want to consider flipping the opt-in, i.e. having detailed errors on by default, and opting out only for ultra-high-perf scenarios.

**Comment by danroot** [comment]
Way out of my league here, but I was curious, so whipped up a [benchmark](https://github.com/danroot/Spike.EnableDetailedErrorsBenchmark).  There probably are some aspects I'm not considering, but does seem to be a perf difference. 

What @ajcvickers suggests seems like it could help.  In pseudocode, instead of:

```
foreach(var row in reader)
foreach(var col in row)
{
   try{
      //...ReadColumn(row,col)
   }catch(Exception ex){
     throw EnrichException(ex,row,col)
   }
}
```

keep state and try-catch the whole block.

```
try{
  Row currentRow = default;
  Column currentColumn = default;
  foreach(var row in reader)
  foreach(var col in row)
  {
       currentRow = row;  currentColumn = col;
       //...ReadColumn(row,col)
  }
}catch(Exception ex){
     throw EnrichException(ex,currentRow,currentColumn)
}
```
... or maybe the equivalent is already being done and the extra allocations for state account for the difference?  An alternative idea: maybe it's not even an EF change that's needed, but a SqlReader change?  It seems like it should already have or be able to get the state needed (column and table names) in the exception path without impacting the "happy path"?

**Comment by roji** [comment]
@danroot can you post the benchmark results so we can see what kind of difference there is? Also, it's usually a good idea to have only the very minimal code you want to test inside your benchmark method - so I'd move the options building out to the [GlobalSetup] method. I'd also enable DbContext pooling to remove the (non-negligible) overhead of instantiating a context instance each time, simulating a more perf-conscious environment.

**Comment by danroot** [comment]
@roji thanks for the tips.  I've updated [the repo](https://github.com/danroot/Spike.EnableDetailedErrorsBenchmark) to move the options building into GlobalSetup and refactor to use pooling (I think?- not sure exactly how best to do that outside of ASP.NET Core DI).  Results were already in the Readme, but I updated them to include memorydiagnoser and re-ran.  Results are below, but now I'm more confused on the results because they seem to indicate it can be slightly faster _with_ EnableDetailedErrors.  Again, grain of salt here - I know just enough to be dangerous about all of this, and the most likely interpretation is I'm doing something wrong in the benchmark. ;)

## Results:
```

BenchmarkDotNet v0.14.0, macOS Sonoma 14.2.1 (23C71) [Darwin 23.2.0]
Apple M1 Max, 1 CPU, 10 logical and 10 physical cores
.NET SDK 9.0.100
  [Host]     : .NET 9.0.0 (9.0.24.52809), Arm64 RyuJIT AdvSIMD
  DefaultJob : .NET 9.0.0 (9.0.24.52809), Arm64 RyuJIT AdvSIMD


```
| Method                           | NumberOfRows | Mean        | Error     | StdDev    | Median      | Gen0      | Gen1     | Gen2     | Allocated   |
|--------------------------------- |------------- |------------:|----------:|----------:|------------:|----------:|---------:|---------:|------------:|
| **QueryWithEnableDetailedErrors**    | **10**           |    **639.0 μs** |  **12.70 μs** |  **17.80 μs** |    **636.4 μs** |    **2.9297** |        **-** |        **-** |    **19.99 KB** |
| QueryWithoutEnableDetailedErrors | 10           |    707.6 μs |  25.00 μs |  72.14 μs |    677.3 μs |    2.9297 |        - |        - |    19.99 KB |
| **QueryWithEnableDetailedErrors**    | **1000**         |  **2,121.8 μs** |  **41.27 μs** |  **36.59 μs** |  **2,122.1 μs** |  **191.4063** |  **70.3125** |        **-** |  **1189.69 KB** |
| QueryWithoutEnableDetailedErrors | 1000         |  2,090.0 μs |  41.21 μs |  76.38 μs |  2,057.2 μs |  191.4063 |  62.5000 |        - |  1189.76 KB |
| **QueryWithEnableDetailedErrors**    | **10000**        | **25,699.0 μs** | **507.43 μs** | **774.91 μs** | **25,419.8 μs** | **2062.5000** | **906.2500** | **281.2500** | **12018.03 KB** |
| QueryWithoutEnableDetailedErrors | 10000        | 25,959.9 μs | 515.26 μs | 786.86 μs | 25,700.2 μs | 2062.5000 | 906.2500 | 281.2500 | 12018.03 KB |

**Comment by roji** [comment]
Thanks. Note that it seems you forgot to clean the database from existing rows, so you're constantly adding on top, skewing the results.

**Comment by danroot** [comment]
@roji TestContainers is recreating the test sql server container for each run.  Essentially for each NumberOfRows parameter value,  get a fresh SQL server instance, create the db, add the correct # of rows, and setup options.  Then run both methods, then dispose the container to throw away the sql instance.  SQL container spinup/teardown doesn't count in benchmark statistics.  I added some logging that confirms this.

**Comment by roji** [comment]
@danroot ah sorry, I missed that you were using a new container instance for each run - all good.

Thanks for the benchmarking, that does show that there's a difference. I think the interesting thing to explore here is whether there's a different possible implementation of the "detailed errors" feature which reduces the overhead... I haven't looked into it, but for example, if there's a try/catch around every field access, it may be better to have a single try/catch for the row, and only update some variable from the field access; that would allow the field access code to get inlined, for example. That's all speculation - we'd need to actually dive into it - but it could be a possible direction.
